core/lib/src/rocket.rs

```diff
@@ -22,38 +25,68 @@ use outcome::Outcome;
 use error::{LaunchError, LaunchErrorKind};
 use fairing::{Fairing, Fairings};
 
+use futures::{Future, Stream};
+use futures::future::{self, FutureResult};
+
 use http::{Method, Status, Header};
 use http::hyper::{self, header};
 use http::uri::Origin;
 
 /// The main `Rocket` type: used to mount routes and catchers and launch the
 /// application.
 pub struct Rocket {
-    crate config: Config,
-    router: Router,
-    default_catchers: HashMap<u16, Catcher>,
-    catchers: HashMap<u16, Catcher>,
-    crate state: Container,
-    fairings: Fairings,
+    crate config: Arc<Config>,
+    router: Arc<Router>,
+    default_catchers: Arc<HashMap<u16, Catcher>>,
+    catchers: Arc<HashMap<u16, Catcher>>,
+    crate state: Arc<Container>,
+    fairings: Arc<Fairings>,
+}
+
+impl<Ctx> hyper::MakeService<Ctx> for Rocket {
```

Comment:
This is one of the key structural differences between our approaches - I used `hyper::service_fn` as in https://github.com/jebrosen/Rocket/commit/27a59ef90dd6e1b978b77be8c0b675b8d547fb08#diff-a9ea99dbf2f3bc38a7d7e4c36a99a288R786. Do you have an opinion on which is better? I prefer `hyper::service_fn` because it's far shorter, but is there a benefit to explicitly specifying the associated types by implementing `MakeService` manually?

Reply:
I had some struggle with `hyper::service_fn` and the idea of using had seemed to me very handy and more robust against further changes in hyper. But that's only a gut feeling. Nevertheless, I think we have to use some kind of cloning of `Arc<SomeRocketStruct>` here.